# NBA Performance

###

### Key Question: Are NBA teams associated with elite individual performance or balanced contributions across players?

**Author:** Imani Niccole Candler

**Tools:** Pandas, Selenium, Python

**Date:** September 7, 2026

In [20]:
import pandas as pd

In [21]:
players_df = pd.read_json("../data/raw/TraditionalPlayerOutput.json")
teams_df = pd.read_json("../data/raw/nba_team_stats.json")


In [22]:
# print the first 5 rows
players_df.head()

,Player Name,Player Link,Team Name,Age,Games Played,Wins,Losses,Points,Three Pt Percentage,Free Throw Percentage,Rebounds,Assists
0,Luka Dončić,https://www.nba.com/stats/player/1629029/,LAL,27,64,43,21,33.5,36.6,78.0,7.7,8.3
1,Shai Gilgeous-Alexander,https://www.nba.com/stats/player/1628983/,OKC,27,65,53,12,31.6,38.0,88.1,4.4,6.5
2,Anthony Edwards,https://www.nba.com/stats/player/1630162/,MIN,24,60,36,24,28.9,39.8,79.6,5.0,3.7
3,Jaylen Brown,https://www.nba.com/stats/player/1627759/,BOS,29,68,44,24,28.7,34.5,79.7,7.0,5.3
4,Tyrese Maxey,https://www.nba.com/stats/player/1630178/,PHI,25,65,36,29,28.7,37.0,89.3,4.2,6.8


In [23]:
teams_df.head()

,Team Name,W,L,WIN%,REB,AST
0,Oklahoma City Thunder,61,16,0.792,44.2,25.4
1,San Antonio Spurs,59,19,0.756,47.1,28.0
2,Detroit Pistons,57,21,0.731,45.6,27.5
3,Boston Celtics,53,25,0.679,46.5,24.6
4,Los Angeles Lakers,50,27,0.649,41.0,25.4


In [24]:
print("player dataset shape:", players_df.shape)
print("team dataset shape:", teams_df.shape)

player dataset shape: (578, 12)
team dataset shape: (30, 6)


In [ ]:
# list the column titles as lists
print("\n player columns:")
print(players_df.columns.tolist())

print("\n team columns:")
print(teams_df.columns.tolist())


 player columns:
['Player Name', 'Player Link', 'Team Name', 'Age', 'Games Played', 'Wins', 'Losses', 'Points', 'Three Pt Percentage', 'Free Throw Percentage', 'Rebounds', 'Assists']

 team columns:
['Team Name', 'W', 'L', 'WIN%', 'REB', 'AST']


In [27]:
print("PLAYER DATA TYPES")
players_df.info()

print("\n TEAM DATA TYPES")
teams_df.info()

PLAYER DATA TYPES
<class 'pandas.DataFrame'>
RangeIndex: 578 entries, 0 to 577
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Player Name            578 non-null    str    
 1   Player Link            578 non-null    str    
 2   Team Name              578 non-null    str    
 3   Age                    578 non-null    int64  
 4   Games Played           578 non-null    int64  
 5   Wins                   578 non-null    int64  
 6   Losses                 578 non-null    int64  
 7   Points                 578 non-null    float64
 8   Three Pt Percentage    578 non-null    float64
 9   Free Throw Percentage  578 non-null    float64
 10  Rebounds               578 non-null    float64
 11  Assists                578 non-null    float64
dtypes: float64(5), int64(4), str(3)
memory usage: 54.3 KB

 TEAM DATA TYPES
<class 'pandas.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 6 co

In [28]:
print("PLAYER MISSING VALUES")
print(players_df.isnull().sum())

print("\nTEAM MISSING VALUES")
print(teams_df.isnull().sum())

PLAYER MISSING VALUES
Player Name              0
Player Link              0
Team Name                0
Age                      0
Games Played             0
Wins                     0
Losses                   0
Points                   0
Three Pt Percentage      0
Free Throw Percentage    0
Rebounds                 0
Assists                  0
dtype: int64

TEAM MISSING VALUES
Team Name    0
W            0
L            0
WIN%         0
REB          0
AST          0
dtype: int64


In [30]:
print("duplicate player rows:", players_df.duplicated().sum())
print("duplicate team rows:", teams_df.duplicated().sum())

duplicate player rows: 0
duplicate team rows: 0


In [31]:
# renaming columns for easier analysis

players_df = players_df.rename(columns={
    "Player Name": "Player_Name",
    "Player Link": "Player_Link",
    "Team Name": "Team",
    "Games Played": "Games_Played",
    "Three Pt Percentage": "Three_Pt_Percentage",
    "Free Throw Percentage": "Free_Throw_Percentage",
})

teams_df = teams_df.rename(columns={
    "Team Name": "Team_Name",
    "W": "Wins",
    "L": "Losses",
    "REB": "Team_Rebounds",
    "AST": "Team_Assists"
})

In [ ]:
# check rename columns
print(players_df.columns.tolist())
print(teams_df.columns.tolist())

['Player_Name', 'Player_Link', 'Team', 'Age', 'Games_Played', 'Wins', 'Losses', 'Points', 'Three_Pt_Percentage', 'Free_Throw_Percentage', 'Rebounds', 'Assists']
['Team_Name', 'Wins', 'Losses', 'WIN%', 'Team_Rebounds', 'Team_Assists']


In [36]:
# taken from nba_team_stats.py
team_map = {
    "ATL": "Atlanta Hawks",
    "BOS": "Boston Celtics",
    "BKN": "Brooklyn Nets",
    "CHA": "Charlotte Hornets",
    "CHI": "Chicago Bulls",
    "CLE": "Cleveland Cavaliers",
    "DAL": "Dallas Mavericks",
    "DEN": "Denver Nuggets",
    "DET": "Detroit Pistons",
    "GSW": "Golden State Warriors",
    "HOU": "Houston Rockets",
    "IND": "Indiana Pacers",
    "LAC": "LA Clippers",
    "LAL": "Los Angeles Lakers",
    "MEM": "Memphis Grizzlies",
    "MIA": "Miami Heat",
    "MIL": "Milwaukee Bucks",
    "MIN": "Minnesota Timberwolves",
    "NOP": "New Orleans Pelicans",
    "NYK": "New York Knicks",
    "OKC": "Oklahoma City Thunder",
    "ORL": "Orlando Magic",
    "PHI": "Philadelphia 76ers",
    "PHX": "Phoenix Suns",
    "POR": "Portland Trail Blazers",
    "SAC": "Sacramento Kings",
    "SAS": "San Antonio Spurs",
    "TOR": "Toronto Raptors",
    "UTA": "Utah Jazz",
    "WAS": "Washington Wizards"
}

In [37]:
# players uses abbreviations while teams uses full team names

players_df["Team_Name"] = players_df["Team"].map(team_map)

In [38]:
players_df[["Team", "Team_Name"]].head(10)

,Team,Team_Name
0,LAL,Los Angeles Lakers
1,OKC,Oklahoma City Thunder
2,MIN,Minnesota Timberwolves
3,BOS,Boston Celtics
4,PHI,Philadelphia 76ers
5,LAC,LA Clippers
6,DEN,Denver Nuggets
7,CLE,Cleveland Cavaliers
8,MIL,Milwaukee Bucks
9,GSW,Golden State Warriors


In [40]:
players_df[["Team", "Team_Name"]].drop_duplicates().sort_values("Team")

,Team,Team_Name
24,ATL,Atlanta Hawks
18,BKN,Brooklyn Nets
3,BOS,Boston Celtics
40,CHA,Charlotte Hornets
74,CHI,Chicago Bulls
7,CLE,Cleveland Cavaliers
35,DAL,Dallas Mavericks
6,DEN,Denver Nuggets
17,DET,Detroit Pistons
9,GSW,Golden State Warriors


# -------------------------------------------------------------------------
# Moving into analysis
### Striving to answer: Are NBA teams associated with elite individual performance or balanced contributions across players?

Let's start with basic team metric summaries:


In [42]:
"""
 groupby = group the 578 player records by NBA team name and abbreviation
 agg = for each player, calculate the following stats
"""

team_player_summary = (
    players_df
    .groupby(["Team", "Team_Name"])
    .agg(
        player_count = ("Player_Name", "count"),
        avg_player_ppg = ("Points", "mean"),
        avg_player_assists = ("Assists", "mean"),
        avg_player_rebounds = ("Rebounds", "mean")
    )
    .reset_index()
)
team_player_summary.head()

,Team,Team_Name,player_count,avg_player_ppg,avg_player_assists,avg_player_rebounds
0,ATL,Atlanta Hawks,20,8.430000,2.060000,3.365000
1,BKN,Brooklyn Nets,20,8.355000,1.980000,3.170000
2,BOS,Boston Celtics,18,8.516667,1.888889,3.883333
3,CHA,Charlotte Hornets,19,8.131579,1.784211,3.468421
4,CHI,Chicago Bulls,19,8.205263,2.005263,3.231579


In [43]:
team_player_summary.shape

(30, 6)

### Finding average player PPG

In [45]:
# now we want to sort teams based on average player ppg (points per game)

team_player_summary.sort_values(
    "avg_player_ppg",
    ascending=False
).head(10)

,Team,Team_Name,player_count,avg_player_ppg,avg_player_assists,avg_player_rebounds
28,UTA,Utah Jazz,21,11.014286,2.871429,4.919048
9,GSW,Golden State Warriors,18,10.322222,2.388889,3.516667
29,WAS,Washington Wizards,22,10.200000,2.422727,4.222727
25,SAC,Sacramento Kings,20,9.825000,2.155000,3.930000
15,MIA,Miami Heat,17,9.623529,2.252941,3.611765
18,NOP,New Orleans Pelicans,17,9.552941,2.170588,3.835294
14,MEM,Memphis Grizzlies,25,9.544000,2.428000,3.932000
24,POR,Portland Trail Blazers,19,9.442105,2.031579,3.463158
22,PHI,Philadelphia 76ers,20,9.395000,2.020000,3.425000
20,OKC,Oklahoma City Thunder,18,9.288889,2.094444,3.561111


### Finding the top scorers for each team

In [ ]:
# sort players by points, the first row is the top scorer for respective team
top_scorers = (
    players_df
    .sort_values("Points", ascending=False)
    .groupby("Team_Name")
    .first()
    .reset_index()
)

# renaming the columns
top_scorers = top_scorers[
    ["Team_Name", "Player_Name", "Points"]
].rename(columns={
    "Player_Name": "Top_Scorer",
    "Points": "Top_Scorer_PPG"
})

# nowwww, sort from highest to lowest player ppg
top_scorers = top_scorers.sort_values(
    "Top_Scorer_PPG",
    ascending=False
).reset_index(drop=True)

top_scorers.head(10)

,Team_Name,Top_Scorer,Top_Scorer_PPG
0,Los Angeles Lakers,Luka Dončić,33.5
1,Oklahoma City Thunder,Shai Gilgeous-Alexander,31.6
2,Minnesota Timberwolves,Anthony Edwards,28.9
3,Philadelphia 76ers,Tyrese Maxey,28.7
4,Boston Celtics,Jaylen Brown,28.7
5,LA Clippers,Kawhi Leonard,28.0
6,Denver Nuggets,Nikola Jokić,27.9
7,Cleveland Cavaliers,Donovan Mitchell,27.7
8,Milwaukee Bucks,Giannis Antetokounmpo,27.6
9,Golden State Warriors,Stephen Curry,27.2


# Core Metric: Scoring Concentration
### Calculate Top-3 Scoring
Identify the three highest-scoring players by PPG:

In [52]:
top_3_scoring = (
    players_df
    .sort_values(
        ["Team_Name", "Points"],
        ascending=[True,False]

    )
    .groupby("Team_Name")
    .head(3)
)
top_3_scoring.head(10)

,Player_Name,Player_Link,Team,Age,Games_Played,Wins,Losses,Points,Three_Pt_Percentage,Free_Throw_Percentage,Rebounds,Assists,Team_Name
24,Jalen Johnson,https://www.nba.com/stats/player/1630552/,ATL,24,69,41,28,22.8,35.1,78.4,10.3,8.0,Atlanta Hawks
37,Nickeil Alexander-Walker,https://www.nba.com/stats/player/1629638/,ATL,27,75,42,33,20.6,39.4,90.0,3.5,3.7,Atlanta Hawks
55,CJ McCollum,https://www.nba.com/stats/player/203468/,ATL,34,73,36,37,18.7,37.1,77.4,3.3,3.8,Atlanta Hawks
3,Jaylen Brown,https://www.nba.com/stats/player/1627759/,BOS,29,68,44,24,28.7,34.5,79.7,7.0,5.3,Boston Celtics
29,Jayson Tatum,https://www.nba.com/stats/player/1628369/,BOS,28,13,11,2,21.4,33.3,82.1,9.9,5.1,Boston Celtics
75,Payton Pritchard,https://www.nba.com/stats/player/1630202/,BOS,28,75,51,24,17.0,37.2,88.6,4.0,5.1,Boston Celtics
18,Michael Porter Jr.,https://www.nba.com/stats/player/1629008/,BKN,27,52,14,38,24.2,36.3,85.9,7.1,3.0,Brooklyn Nets
148,Noah Clowney,https://www.nba.com/stats/player/1641730/,BKN,21,66,17,49,12.3,32.9,80.4,4.1,1.6,Brooklyn Nets
164,Nic Claxton,https://www.nba.com/stats/player/1629651/,BKN,26,69,17,52,11.7,15.8,61.6,6.9,3.7,Brooklyn Nets
40,Brandon Miller,https://www.nba.com/stats/player/1641706/,CHA,23,61,39,22,20.4,39.2,89.2,5.0,3.4,Charlotte Hornets


In [54]:
# add top-3 ppg values

top_3_summary = (
    top_3_scoring
    .groupby("Team_Name")
    .agg(
        Top_3_PPG = ("Points", "sum")
    )
    .reset_index()
)

top_3_summary = (
    top_3_summary
    .sort_values("Top_3_PPG", ascending=False)
    .reset_index(drop=True)
)

top_3_summary.head(10)

,Team_Name,Top_3_PPG
0,Los Angeles Lakers,77.4
1,Philadelphia 76ers,73.1
2,Utah Jazz,69.7
3,Denver Nuggets,69.7
4,Cleveland Cavaliers,69.4
5,Boston Celtics,67.1
6,Oklahoma City Thunder,65.6
7,LA Clippers,65.4
8,Minnesota Timberwolves,64.8
9,Golden State Warriors,64.5


### Calculating Scoring Concentration
### Formula: ((Top 3 combined PPG) / (Top 10 combined PPG)) x 100
**Meaning:** A higher percentage means scoring is more concentrated among the top 3 scorers. A lower percentage means scoring is more concentrated acorss the top 10 scorers

### Get the top 10 players

In [55]:
top_10_scoring = (
    players_df
    .sort_values(
        ["Team_Name", "Points"],
        ascending=[True, False]
    )
    .groupby("Team_Name")
    .head(10)
)

In [56]:
top_10_summary = (
    top_10_scoring
    .groupby("Team_Name")
    .agg(
        Top_10_PPG = ("Points", "sum")
    )
    .reset_index()
)

### Combine scoring summaries

In [60]:
# combining 3 dataframes into 1

scoring_summary = (
    top_scorers
    .merge(top_3_summary, on="Team_Name")
    .merge(top_10_summary, on="Team_Name")
)

scoring_summary.head()

,Team_Name,Top_Scorer,Top_Scorer_PPG,Top_3_PPG,Top_10_PPG
0,Los Angeles Lakers,Luka Dončić,33.5,77.4,139.9
1,Oklahoma City Thunder,Shai Gilgeous-Alexander,31.6,65.6,134.7
2,Minnesota Timberwolves,Anthony Edwards,28.9,64.8,134.9
3,Philadelphia 76ers,Tyrese Maxey,28.7,73.1,147.4
4,Boston Celtics,Jaylen Brown,28.7,67.1,136.7


# -------------------------------------------------------------------------
# Calculating Scoring Concentration!
**Formula:** ((Top 3 combined PPG) / (Top 10 combined PPG)) x 100

In [61]:
scoring_summary["Scoring_Concentration"] = (
    scoring_summary["Top_3_PPG"] /
    scoring_summary["Top_10_PPG"]
) * 100

scoring_summary["Scoring_Concentration"] = (
    scoring_summary["Scoring_Concentration"].round(2)
)

# NOTE!
### Don't confuse scoring concentration as the percentage of the team's actual points, but the sum of individual PPG avgs.
Higher % --> scoring is more concentrated among top 3 players  
Lower % --> scoring is more distributed among top 10 players

In [62]:
scoring_summary.sort_values(
    "Scoring_Concentration",
    ascending=False
).head(10)

,Team_Name,Top_Scorer,Top_Scorer_PPG,Top_3_PPG,Top_10_PPG,Scoring_Concentration
0,Los Angeles Lakers,Luka Dončić,33.5,77.4,139.9,55.33
12,Houston Rockets,Kevin Durant,25.8,64.3,123.4,52.11
5,LA Clippers,Kawhi Leonard,28.0,65.4,130.2,50.23
3,Philadelphia 76ers,Tyrese Maxey,28.7,73.1,147.4,49.59
4,Boston Celtics,Jaylen Brown,28.7,67.1,136.7,49.09
1,Oklahoma City Thunder,Shai Gilgeous-Alexander,31.6,65.6,134.7,48.70
6,Denver Nuggets,Nikola Jokić,27.9,69.7,144.4,48.27
2,Minnesota Timberwolves,Anthony Edwards,28.9,64.8,134.9,48.04
20,Orlando Magic,Paolo Banchero,22.2,63.5,133.7,47.49
11,New York Knicks,Jalen Brunson,26.0,63.0,133.1,47.33


### Most Concentrated

In [63]:
# looking at the two extremes

scoring_summary.nlargest(
    5,
    "Scoring_Concentration"
)[
    ["Team_Name", "Top_Scorer", "Top_Scorer_PPG",
     "Top_3_PPG", "Top_10_PPG", "Scoring_Concentration"]
]

,Team_Name,Top_Scorer,Top_Scorer_PPG,Top_3_PPG,Top_10_PPG,Scoring_Concentration
0,Los Angeles Lakers,Luka Dončić,33.5,77.4,139.9,55.33
12,Houston Rockets,Kevin Durant,25.8,64.3,123.4,52.11
5,LA Clippers,Kawhi Leonard,28.0,65.4,130.2,50.23
3,Philadelphia 76ers,Tyrese Maxey,28.7,73.1,147.4,49.59
4,Boston Celtics,Jaylen Brown,28.7,67.1,136.7,49.09


### Most Balanced

In [64]:
scoring_summary.nsmallest(
    5,
    "Scoring_Concentration"
)[
   ["Team_Name", "Top_Scorer", "Top_Scorer_PPG",
    "Top_3_PPG", "Top_10_PPG", "Scoring_Concentration"]
]

,Team_Name,Top_Scorer,Top_Scorer_PPG,Top_3_PPG,Top_10_PPG,Scoring_Concentration
28,Sacramento Kings,Zach LaVine,19.2,53.6,142.0,37.75
27,Memphis Grizzlies,Ty Jerome,19.7,53.2,138.5,38.41
29,Chicago Bulls,Josh Giddey,17.0,48.5,120.5,40.25
26,Washington Wizards,Anthony Davis,20.4,54.6,134.4,40.62
24,Dallas Mavericks,Cooper Flagg,20.8,50.3,123.8,40.63


### Merge with team performance

In [65]:
nba_analysis = teams_df.merge(
    scoring_summary,
    on="Team_Name",
    how="left"
)

nba_analysis.head()

,Team_Name,Wins,Losses,WIN%,Team_Rebounds,Team_Assists,Top_Scorer,Top_Scorer_PPG,Top_3_PPG,Top_10_PPG,Scoring_Concentration
0,Oklahoma City Thunder,61,16,0.792,44.2,25.4,Shai Gilgeous-Alexander,31.6,65.6,134.7,48.70
1,San Antonio Spurs,59,19,0.756,47.1,28.0,Victor Wembanyama,24.9,60.1,130.3,46.12
2,Detroit Pistons,57,21,0.731,45.6,27.5,Cade Cunningham,24.5,57.1,123.9,46.09
3,Boston Celtics,53,25,0.679,46.5,24.6,Jaylen Brown,28.7,67.1,136.7,49.09
4,Los Angeles Lakers,50,27,0.649,41.0,25.4,Luka Dončić,33.5,77.4,139.9,55.33


In [ ]:
# calculate the correlation coefficient between 3 variables
nba_analysis[
    ["WIN%", "Scoring_Concentration", "Top_Scorer_PPG"]
].corr()

,WIN%,Scoring_Concentration,Top_Scorer_PPG
WIN%,1.000000,0.716467,0.555551
Scoring_Concentration,0.716467,1.000000,0.794448
Top_Scorer_PPG,0.555551,0.794448,1.000000


### What the Data Says ~ Correlation Coefficient 
### Focusing on the relationship between Win Percentage and Scoring Concentration = 0.716467

**Positive:** higher-performing teams tend to have more concentrated top-3 scoring.
**Negative:** higher-performing teams tend to have more distributed scoring.
**Near Zero:** scoring concentratoin doesn't have a strong relationship with winning

In [83]:
win_concentration_corr = nba_analysis[
    "WIN%"
].corr(
    nba_analysis["Scoring_Concentration"]
)

win_top_corr = nba_analysis[
    "WIN%"
].corr(
    nba_analysis["Top_Scorer_PPG"]
)

print(
    f"Win % vs. Scoring Concentration: "
    f"{win_concentration_corr:.3f}"
)


print(
    f"Win % vs. Top Scorer PPG: "
    f"{win_top_corr:.3f}"
)

Win % vs. Scoring Concentration: 0.716
Win % vs. Top Scorer PPG: 0.556


### Scoring Findings

Scoring concentration has a **strong positive correlation** with win percentage (r = 0.716), while top-scorer PPG has a moderate positive correlation with win percentage (r = 0.556).

Through this dataset, team success was more strongly associated with scoring among the top three players than the team's leading scorer alone.

These relationships are correlational, not causational.

# Does team-wide balance show up in areas *other than scoring?*

### Also ask, are winning teams generally better at assisting and rebounding?

In [84]:
team_stat_corr = nba_analysis[
    ["WIN%", "Team_Assists", "Team_Rebounds"]
].corr()

team_stat_corr

,WIN%,Team_Assists,Team_Rebounds
WIN%,1.000000,0.070961,0.511276
Team_Assists,0.070961,1.000000,0.139202
Team_Rebounds,0.511276,0.139202,1.000000


### Assist Concentration
**Formula:** Assist Concentration = ((Top 3 players' APG) / (Top 10 players' APG)) x 100  
Higher = assists are concentrated among a few players  
Lower = assists are balanced among top 10.

In [85]:
top_3_assists = (
    players_df
    .sort_values(
        ["Team_Name", "Assists"],
        ascending=[True, False]
    )
    .groupby("Team_Name")
    .head(3)
)

top_10_assists = (
    players_df
    .sort_values(
        ["Team_Name", "Assists"],
        ascending=[True, False]
    )
    .groupby("Team_Name")
    .head(10)
)


In [86]:
assist_summary = (
    top_3_assists
    .groupby("Team_Name")
    .agg(Top_3_APG = ("Assists", "sum"))
    .reset_index()
)

top_10_assist_summary = (
    top_10_assists
    .groupby("Team_Name")
    .agg(Top_10_APG = ("Assists", "sum"))
    .reset_index()
)


In [87]:
# merge

assist_summary = assist_summary.merge(
    top_10_assist_summary,
    on = "Team_Name")


In [88]:
assist_summary["Assist_Concentration"] = (
    assist_summary["Top_3_APG"] /
    assist_summary["Top_10_APG"]
) * 100

assist_summary["Assist_Concentration"] = (
    assist_summary["Assist_Concentration"].round(2)
)

assist_summary.sort_values(
    "Assist_Concentration",
    ascending=False
).head(10)

,Team_Name,Top_3_APG,Top_10_APG,Assist_Concentration
13,Los Angeles Lakers,20.7,31.5,65.71
10,Houston Rockets,16.2,27.6,58.70
7,Denver Nuggets,20.9,36.8,56.79
4,Chicago Bulls,17.8,32.0,55.62
26,San Antonio Spurs,17.4,31.3,55.59
16,Milwaukee Bucks,18.4,34.1,53.96
27,Toronto Raptors,17.3,32.9,52.58
0,Atlanta Hawks,17.7,33.8,52.37
1,Boston Celtics,15.9,30.6,51.96
24,Portland Trail Blazers,16.6,32.2,51.55


In [90]:
nba_analysis = nba_analysis.merge(
    assist_summary,
    on="Team_Name",
    how="left"
)

In [91]:
nba_analysis[
    ["WIN%", "Team_Assists", "Assist_Concentration"]
].corr()

,WIN%,Team_Assists,Assist_Concentration
WIN%,1.000000,0.070961,0.480614
Team_Assists,0.070961,1.000000,0.079890
Assist_Concentration,0.480614,0.079890,1.000000


### What are the numbers saying

Moderate positive relationship between assist concentration and win percentage (r = 0.481)

### Testing Rebounds

In [92]:
top_3_rebounds = (
    players_df
    .sort_values(
        ["Team_Name", "Rebounds"],
        ascending = [True, False]
    )
    .groupby("Team_Name")
    .head(3)
)

top_10_rebounds = (
    players_df
    .sort_values(
        ["Team_Name", "Rebounds"],
        ascending = [True, False]
    )
    .groupby("Team_Name")
    .head(10)
)
        

In [93]:
rebound_summary = (
    top_3_rebounds
    .groupby("Team_Name")
    .agg(Top_3_RPG = ("Rebounds", "sum"))
    .reset_index()
)

top_10_rebound_summary = (
    top_10_rebounds
    .groupby("Team_Name")
    .agg(Top_10_RPG = ("Rebounds", "sum"))
    .reset_index()
)

rebound_summary = rebound_summary.merge(
    top_10_rebound_summary,
    on = "Team_Name"
)


In [94]:
rebound_summary["Rebound_Concentration"] = (
    rebound_summary["Top_3_RPG"] / 
    rebound_summary["Top_10_RPG"]
) * 100

rebound_summary["Rebound_Concentration"] = (
    rebound_summary["Rebound_Concentration"].round(2)
)

In [96]:
# merge into dataset

nba_analysis = nba_analysis.merge(
    rebound_summary,
    on = "Team_Name",
    how = "left"
)

In [99]:
nba_analysis[
    ["WIN%", "Team_Rebounds", "Rebound_Concentration"]
].corr()

,WIN%,Team_Rebounds,Rebound_Concentration
WIN%,1.000000,0.511276,0.353522
Team_Rebounds,0.511276,1.000000,0.248659
Rebound_Concentration,0.353522,0.248659,1.000000


# -------------------------------------------------------------------------
# Final Findings

In [101]:
results = pd.DataFrame({
    "Metric": [
        "Scoring Concentration",
        "Top Scorer PPG",
        "Team Rebounds",
        "Assist Concentration",
        "Rebound Concentration",
        "Team Assists"
    ],
    "Correlation_with_WIN%": [
        0.716467,
        0.555551,
        0.511276,
        0.480614,
        0.353522,
        0.070961
    ]
})

results["Correlation_with_WIN%"] = (
    results["Correlation_with_WIN%"].round(3)
)

results

,Metric,Correlation_with_WIN%
0,Scoring Concentration,0.716
1,Top Scorer PPG,0.556
2,Team Rebounds,0.511
3,Assist Concentration,0.481
4,Rebound Concentration,0.354
5,Team Assists,0.071


# Key Question: Are team rankings in the NBA driven by elite individual players or balanced team wide statistical contributors?
### Key Findings - Stars *do* run the league!
~ Teams with scoring concentrated at the top three players tended to have a higher win percentage (r = 0.716).  
~ A top-3 scoring concentration is more closely associated with winning than one elite player. The Top Scorer's PPG was r = 0.556.  

### Secondary Question: Are winning teams generally better at assisting and rebounding?
~ Assist Concentration (r = 0.481) has a stronger relationship with win percentage than Rebound Concentration (r = 0.354) and win percentage.  
~ In relation to win percentage, the results lean more towards assists  among three players (r = 0.481) than the overall team (r = 0.071).  
~ In relation to win percentage, the results lean more towards rebounds in overall team production (r = 0.511) than concentration among three players (r = 0.354).   

*NOTE:* These findings describe associations within the dataset and don't establish that player production causes NBA teams to win.

# Limitations
~ Analysis is limited to data from the 2025-2026 NBA regular season.  
~ Study excludes postseason performance, which may yield different statistical trends.  
~ Variables are concentrated on offensive output; defensive impact was not evaluated.

In [103]:
name_to_abbr = {name: abbr for abbr, name in team_map.items()}

nba_analysis["Team"] = nba_analysis["Team_Name"].map(name_to_abbr)

In [104]:
nba_analysis[
    ["Team", "Team_Name", "WIN%", "Scoring_Concentration"]
].head()

,Team,Team_Name,WIN%,Scoring_Concentration
0,OKC,Oklahoma City Thunder,0.792,48.70
1,SAS,San Antonio Spurs,0.756,46.12
2,DET,Detroit Pistons,0.731,46.09
3,BOS,Boston Celtics,0.679,49.09
4,LAL,Los Angeles Lakers,0.649,55.33


In [106]:
nba_analysis.to_csv(
    "../data/processed/nba_team_analysis.csv",
    index=False
)

In [107]:
print(nba_analysis.shape)
print(nba_analysis.columns.tolist())

(30, 18)
['Team_Name', 'Wins', 'Losses', 'WIN%', 'Team_Rebounds', 'Team_Assists', 'Top_Scorer', 'Top_Scorer_PPG', 'Top_3_PPG', 'Top_10_PPG', 'Scoring_Concentration', 'Top_3_APG', 'Top_10_APG', 'Assist_Concentration', 'Top_3_RPG', 'Top_10_RPG', 'Rebound_Concentration', 'Team']
